users/vieting/experiments/switchboard/ctc/feat/baseline_args.py

```diff
@@ -237,6 +253,40 @@ def get_returnn_config(
         # network["source"] = specaug_layer_jingjing(in_layer=["features"])
         pass
 
+    if audio_perturbation and recognition:
+        if preemphasis_perturbation:
+            for layer in feature_net["subnetwork"]:
+                layer_config = feature_net["subnetwork"][layer]
+                if layer_config.get("class") != "variable":
+                    if feature_net["subnetwork"][layer].get("from") == "data":
+                        feature_net["subnetwork"][layer]["from"] = "preemphasis"
+            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=0.97).get_as_subnetwork()
+        elif codec_perturbation:
+            for layer in feature_net["subnetwork"]:
+                layer_config = feature_net["subnetwork"][layer]
+                if layer_config.get("class") != "variable":
+                    if feature_net["subnetwork"][layer].get("from") == "data":
+                        feature_net["subnetwork"][layer]["from"] = "codec"
+            feature_net["subnetwork"]["codec"] = {
+                "class": "subnetwork",
+                "from": ["data"],
+                "subnetwork": {
+                    "safe_input": {"class": "eval", "from": "data", "eval": "tf.clip_by_value(source(0), -1.0, 1.0)"},
```

Comment:
Why is this needed?

Reply:
To ensure the data are in the correct range for the mu law codec.

Reply:
Oh I see, there used to be an assertion in the perturbation code which was removed. See my comment there.

I think an assertion is much better suited. With this case, you'll not notice if the data is not in the correct range and just clip it which could seriously hamper the performance. Can you do an assertion here as well?

Reply:
I dont think I can. There is no way for me to check the values of the data before I configure the network and later it would be catched by the readded check in perturbation. I will just remove the layer

Reply:
No the check is not active during recognition. Can't you somehow assert this in tf in that layer?

Reply:
Done.